sstables/mx/reader.cc

```diff
@@ -1781,13 +1797,16 @@ class mx_sstable_full_scan_reader : public mp_row_consumer_reader_mx {
     }
     virtual future<> fill_buffer() override {
         if (_end_of_stream) {
-            return make_ready_future<>();
+            co_return;
+        }
+        if (!is_initialized()) {
+            co_await maybe_initialize();
         }
         if (_context->eof()) {
             _end_of_stream = true;
-            return make_ready_future<>();
+            co_return;
         }
-        return _context->consume_input();
+        co_await _context->consume_input();
```

Comment:
In general `fill_buffer()` is hot path for readers, so we don't use co-routines here, because they add an unconditional allocation.
One could argue that this doesn't apply to `full_scan_reader`, because it is not used in any performance-sensitive code-paths, but I think it is worth keeping all reader's `fill_buffer()` coroutine-free.
There is a simple workaround to keep the code relatively straightforward:
```c++
future<> fill_buffer() {
   if (!is_initialized()) {
       return initialize().then([this] { return fill_buffer(); };
    }
    // rest of the code for `fill_buffer()`.
```

This trick allows for saving an indentation level, at the expense of a recursion and this is considered quite a standard practice for such cases.

Reply:
Note: it is fine to use coroutines in `initialize()` because it is called once in the lifetime of the reader.

Reply:
Ack, I wasn't aware of the allocation overhead. I just wanted to make the code more readable.